### PR TITLE
Use PMCD returned by mappinganalysis to build minimal graph for query

### LINKED ISSUE #2346: Feature request: Improve load time of Query
### Similar requests

- [X] I have searched and found no existing similar requests

### How are you using Legend?

Legend Query

### What problems are you trying to solve?

Currently loading a query takes a lot of time for queries created from huge projects. This task focuses on improving the load time by using caching.

### Describe the solution you would like to see implemented

- [ ] Add metrics to query load time
       

### Describe alternatives you have considered

_No response_

### Documentation, Design, Adoption, Migration Strategy

_No response_

### Contribution

- [ ] I would like to work on this feature

### PR DESCRIPTION
**Pull request by gayathrir11**
## Summary

Depends on finos/legend-engine#2152
closes #2346 

## How did you test this change?

- [ ] Test(s) added
- [ ] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

### CODE REVIEW

File: packages/legend-application-query/src/stores/QueryEditorStore.ts [hunk 1]
```diff
@@ -485,7 +489,9 @@ export abstract class QueryEditorStore {
       );
 
       yield this.setUpEditorState();
-      yield flowResult(this.buildGraph());
+      if (this.requiresGraphBuilding()) {
+        yield flowResult(this.buildGraph());
```
Comment (akphi) [inline packages/legend-application-query/src/stores/QueryEditorStore.ts:493]: with all of this, why don't we just simply **override** `buildGraph()`? the code change is getting a bit wild :)

In the future, if building graph from small cache is the way we go about in majority of the cases, we can introduce a new function to get the cache and default `buildGraph()` to make use of that function

File: packages/legend-application-query/src/stores/LegendQueryApplicationPlugin.ts [hunk 2]
```diff
@@ -52,10 +51,13 @@ export type QuerySetupActionConfiguration = {
 };
 
 export type ExistingQueryEditorStateBuilder = (
-  query: Query,
   editorStore: ExistingQueryEditorStore,
 ) => Promise<QueryBuilderState | undefined>;
 
+export type ExistingQueryGraphBuilderChecker = (
```
Comment (akphi) [inline packages/legend-application-query/src/stores/LegendQueryApplicationPlugin.ts:57]: see my note in `LegendQueryEditor`, probably don't need this

File: packages/legend-extension-dsl-data-space/src/graph-manager/protocol/pure/DSL_DataSpace_PureGraphManagerExtension.ts [hunk 3]
```diff
@@ -34,10 +39,23 @@ export abstract class DSL_DataSpace_PureGraphManagerExtension extends AbstractPu
...
-    cacheRetriever: () => Promise<PlainObject<DataSpaceAnalysisResult>>,
+  abstract analyzeDataSpaceCoverage(
+    dataSpacePath: string,
+    entitiesRetriever: () => Promise<Entity[]>,
+    cacheRetriever?: () => Promise<PlainObject<DataSpaceAnalysisResult>>,
     actionState?: ActionState,
-  ): Promise<DataSpaceAnalysisResult | undefined>;
+  ): Promise<V1_DataSpaceAnalysisResult>;
+
+  abstract buildDataSpaceAnalytics(
```
Comment (akphi) [inline packages/legend-extension-dsl-data-space/src/graph-manager/protocol/pure/DSL_DataSpace_PureGraphManagerExtension.ts:49]: it's fine to expose this method, by why
```ts
analytics:
      | PlainObject<V1_DataSpaceAnalysisResult>
      | V1_DataSpaceAnalysisResult,
```

Comment (akphi) [reply packages/legend-extension-dsl-data-space/src/graph-manager/protocol/pure/DSL_DataSpace_PureGraphManagerExtension.ts:49]: also, do you want to do this, so we don't have to build the whole `dataSpaceAnalytics`? Else, why can't we just use `analyzeDataSpaceCoverage()`?

**Review by akphi — CHANGES_REQUESTED** [review]

File: packages/legend-graph/src/graph-manager/action/analytics/MappingModelCoverageAnalysis.ts [hunk 4]
```diff
@@ -22,10 +23,12 @@ export class MappedEntity {
   readonly __PROPERTIES_INDEX = new Map<string, MappedProperty>();
 
   path: string;
+  classPath: string;
```
[the remark this answers is not in the document]
Comment (gayathrir11) [reply packages/legend-graph/src/graph-manager/action/analytics/MappingModelCoverageAnalysis.ts:26]: Yes. I have linked it in description

File: packages/legend-application-query/src/stores/QueryEditorStore.ts (hunk 1 above)
Comment (gayathrir11) [reply packages/legend-application-query/src/stores/QueryEditorStore.ts:493]: I do override buildGraph() for DataSpaceCreatorStore but for ExistingQueryEditorStore I need to use it via extensions.
But i guess creating a new function would be much cleaner

File: packages/legend-extension-dsl-data-space/src/graph-manager/protocol/pure/DSL_DataSpace_PureGraphManagerExtension.ts (hunk 3 above)
Comment (gayathrir11) [reply packages/legend-extension-dsl-data-space/src/graph-manager/protocol/pure/DSL_DataSpace_PureGraphManagerExtension.ts:49]: I think I can just use analyzeDataSpaceCoverage without exposing this

**Review by akphi — APPROVED** [review]

**Review by YannanGao-gs — APPROVED** [review]

**Comment by YannanGao-gs** [comment]
updated/cleaned by  https://github.com/finos/legend-studio/pull/3488

**Comment by akphi** [comment]
superseded by https://github.com/finos/legend-studio/pull/3488

**Pull request closed without merging**